ext/duckdb/result.c

```diff
@@ -59,373 +36,187 @@ static size_t memsize(const void *p) {
+    duckdb_type type_id = duckdb_get_type_id(ty);
+    void* vector_data = duckdb_vector_get_data(vector);
+
+    switch(type_id) {
+        case DUCKDB_TYPE_INVALID:
+            return Qnil;
+        case DUCKDB_TYPE_BOOLEAN:
+            return (((bool*) vector_data)[row_idx]) ? Qtrue : Qfalse;
+        case DUCKDB_TYPE_TINYINT:
+            return LL2NUM(((int8_t *) vector_data)[row_idx]);
+        case DUCKDB_TYPE_SMALLINT:
+            return LL2NUM(((int16_t *) vector_data)[row_idx]);
+        case DUCKDB_TYPE_INTEGER:
+            return LL2NUM(((int32_t *) vector_data)[row_idx]);
+        case DUCKDB_TYPE_BIGINT:
+            return LL2NUM(((int64_t *) vector_data)[row_idx]);
+        case DUCKDB_TYPE_HUGEINT:
+            // not done
+        case DUCKDB_TYPE_FLOAT:
+            return rb_float_new((((float *) vector_data)[row_idx]));
+        case DUCKDB_TYPE_DOUBLE:
+            return rb_float_new((((double *) vector_data)[row_idx]));
+        case DUCKDB_TYPE_DATE:
+            return vector_date(vector_data, row_idx);
+        case DUCKDB_TYPE_TIMESTAMP:
+            return vector_timestamp(vector_data, row_idx);
+        case DUCKDB_TYPE_INTERVAL:
+            // not done
+        case DUCKDB_TYPE_VARCHAR:
+        case DUCKDB_TYPE_BLOB:
+            return vector_blob(vector_data, row_idx);
+        case DUCKDB_TYPE_DECIMAL:
+            return vector_decimal(ty, vector_data, row_idx);
+        case DUCKDB_TYPE_LIST:
+            return vector_list(vector, row_idx);
+        case DUCKDB_TYPE_MAP:
+        case DUCKDB_TYPE_STRUCT:
+            return vector_struct(ty, vector, row_idx);
+        case DUCKDB_TYPE_UUID:
+            // this is the same as the hugeint, we just flip the sign bit
+        default:
+            return Qnil;
     }
-    return Qnil;
 }
 
-static VALUE duckdb_result__to_blob(VALUE oDuckDBResult, VALUE row_idx, VALUE col_idx) {
-    rubyDuckDBResult *ctx;
-    TypedData_Get_Struct(oDuckDBResult, rubyDuckDBResult, &result_data_type, ctx);
+const int stringInlineLength = 12;
 
-    return to_ruby_obj_blob(&(ctx->result), NUM2LL(col_idx), NUM2LL(row_idx));
+static VALUE vector_timestamp(void* vector_data, idx_t row_idx) {
+    duckdb_timestamp ts = ((duckdb_timestamp *)vector_data)[row_idx];
+    return rb_time_new(ts.micros, 0);
 }
 
-static VALUE duckdb_result__enum_internal_type(VALUE oDuckDBResult, VALUE col_idx) {
-    rubyDuckDBResult *ctx;
-    VALUE type = Qnil;
-    duckdb_logical_type logical_type;
-
-    TypedData_Get_Struct(oDuckDBResult, rubyDuckDBResult, &result_data_type, ctx);
-    logical_type = duckdb_column_logical_type(&(ctx->result), NUM2LL(col_idx));
-    if (logical_type) {
-        type = LL2NUM(duckdb_enum_internal_type(logical_type));
+static VALUE vector_blob(void* vector_data, idx_t row_idx) {
+    duckdb_string_t s = (((duckdb_string_t *)vector_data)[row_idx]);
+    if(duckdb_string_is_inlined(s)) {
+        return rb_str_new2(s.value.inlined.inlined);
```

Comment:
Blob data is including `\0`. So you should use `rb_str_new` instead of `rb_str_new2`.

For example, if the blob is `"\0\1\2\3\4\5".force_encoding('ASCII-8BIT')"`, then `rb_str_new2` does not work well.

Reply:
Good call.